src/cache-manager.js

```diff
@@ -106,62 +129,68 @@ class CacheManager {
     assert(rawUrl);
     this.debug(`putting ${rawUrl}`);
 
-    let m = meter();
-    m.on('error', err => {
-      this.debug(`error from stream-meter ${err.stack || err}`);
-    });
+    // Tell others that we're working on this url
+    await this.insertCacheEntry(rawUrl, 'pending', this.cacheTTL);
 
-    this.debug(`creating read stream for ${rawUrl}`);
-    let inputUrlInfo = await this.createUrlReadStream(rawUrl);
-    this.debug(`created read stream for ${rawUrl}`);
-
-    let inputStream = inputUrlInfo.stream;
-
-    // We need the following pieces of information in the service-specific
-    // implementations
-    let contentType = inputUrlInfo.meta.headers[inputUrlInfo.meta.caseless.has('content-type')];
-    contentType = contentType || 'application/octet-stream';
-    let upstreamEtag = inputUrlInfo.meta.headers[inputUrlInfo.meta.caseless.has('etag')];
-    upstreamEtag = upstreamEtag || '';
-    let contentEncoding = inputUrlInfo.meta.headers[inputUrlInfo.meta.caseless.has('content-encoding')];
-    let contentDisposition = inputUrlInfo.meta.headers[inputUrlInfo.meta.caseless.has('content-disposition')];
-
-    let headers = {
-      'Content-Type': contentType,
-      'Content-Disposition': contentDisposition,
-      'Content-Encoding': contentEncoding,
-    };
+    // Basically, any error here should do the same thing: pring the exception
+    // in our logs then set the cache entry to status === 'error'
+    try {
+      let m = meter();
+      m.on('error', err => {
+        this.debug(`error from stream-meter ${err.stack || err}`);
+      });
 
-    let storageMetadata = {
-      'upstream-etag': upstreamEtag,
-      url: rawUrl,
-      stored: new Date().toISOString(),
-      addresses: JSON.stringify(inputUrlInfo.addresses),
-    };
+      this.debug(`creating read stream for ${rawUrl}`);
+      let inputUrlInfo = await this.createUrlReadStream(rawUrl);
+      this.debug(`created read stream for ${rawUrl}`);
+
+      let inputStream = inputUrlInfo.stream;
+
+      // We need the following pieces of information in the service-specific
+      // implementations
+      let contentType = inputUrlInfo.meta.headers[inputUrlInfo.meta.caseless.has('content-type')];
+      contentType = contentType || 'application/octet-stream';
+      let upstreamEtag = inputUrlInfo.meta.headers[inputUrlInfo.meta.caseless.has('etag')];
+      upstreamEtag = upstreamEtag || '';
+      let contentEncoding = inputUrlInfo.meta.headers[inputUrlInfo.meta.caseless.has('content-encoding')];
+      let contentDisposition = inputUrlInfo.meta.headers[inputUrlInfo.meta.caseless.has('content-disposition')];
+      let contentMD5 = inputUrlInfo.meta.headers[inputUrlInfo.meta.caseless.has('content-md5')];
+
+      let headers = {
+        'Content-Type': contentType,
+        'Content-Disposition': contentDisposition,
+        'Content-Encoding': contentEncoding,
+        'Content-MD5': contentMD5,
```

Comment:
In node it's custom to do headers in lowercase... So why not follow that convention?
Infact since you are picking them from the `inputUrlInfo.meta` this could just be:
`let headers = _.pick(inputUrlInfo.meta.headers, 'content-md5', 'content-type', 'content-encoding', 'content-disposition');`

Or am I missing something?